### DISCUSSION TITLE
Alternate is failing w/ directory outside .git root. How is this supposed to work?

### DISCUSSION THREAD

**Discussion by jemiahw** [go-git/go-git#1125] [Q&A]
I have a git repo in `/builds/default-i-044c26eb54cb1e663-7/foo/myproject`. 

That repo has an alternate that points to an absolute directory outside of the repo. The `.git/objects/info/alternates` file points to `/mirrors/https---github-com-foo-myproject/objects` (note that this is an absolute path, outside the current repo).

Following the code, I see it drops into the [DotGit.Alternates()](https://github.com/go-git/go-git/blob/master/storage/filesystem/dotgit/dotgit.go#L1147) method to handle the alternate. Since it is an absolute path, it drops into the block on line [1176 here](https://github.com/go-git/go-git/blob/master/storage/filesystem/dotgit/dotgit.go#L1176). The fs type is a ChrootHelper, so it drops into the block to convert this to a relative path. It converts it to `../../../../../mirrors/https---github-com-foo-myproject/objects`.

Then it tries to call `fs.Stat("../../../../../mirrors/https---github-com-foo-myproject/objects")` on that path on [line 1196](https://github.com/go-git/go-git/blob/master/storage/filesystem/dotgit/dotgit.go#L1196C3-L1198), which fails with an error `chroot boundary crossed`.

So it looks like this doesn't work if the alternate directory is outside the `.git` directory since that would always cross the chroot boundary. But isn't the point of the alternate that the objects exist somewhere else? I don't see how this ever works, but it must work somehow. Do I need to do some other initialization other than PlainOpen?

**Comment by pjbgf** [comment]
@jemiahw please note that you can set a different FS to the one used for the worktree and dotgit. Refer to the tests for [dotgit.NewWithOptions](https://github.com/pjbgf/go-git/blob/8b47ceb1aa854f3c3bfa1c347157a04324fcd51e/storage/filesystem/dotgit/dotgit_test.go#L886) for more details.

**Reply by jemiahw** [reply]
I got one step further with the alternate filesystem. I'm running into a different issue now, when I call Status() it shows all files as deleted. 

I suspect this is something to do with how I'm opening things. The files exist under the `repoRoot` directory, but when I open the first filesystem I had to explicitly open the `repoRoot + ".git"` directory. So maybe that filesystem doesn't see them because it's one directory level down? And the alternate is in another directory tree entirely, which has the `object` files but not the files themselves.  I'll post a follow up if I make additional progress, but for now, this is where I am.

```	
	fs := osfs.New(repoRoot + ".git")
	storer := filesystem.NewStorageWithOptions(fs, cache.NewObjectLRUDefault(), filesystem.Options{
		AlternatesFS: osfs.New("/"),
	})
	repo, err := git.Open(storer, fs)
```

**Reply by pjbgf** [reply]
The default `osfs` uses the [ChrootOS](https://github.com/go-git/go-billy/blob/master/osfs/os_chroot.go#L14-L23), so whatever dir you provide to it will become `/` within its fs. Potentially that is what is leading to some of the issues you are experiencing here.

For alternate FS, it may be easier for you to use [BoundOS](https://github.com/go-git/go-billy/blob/master/osfs/os_bound.go#L32-L42) instead, so that you have the same representation as the actual OS, but the operations are bound to a specific dir.
On that specific case, this could look like:

```golang
	fs := osfs.New("/git/some-repo.git", osfs.WithBoundOS())
	storer := filesystem.NewStorageWithOptions(fs, cache.NewObjectLRUDefault(), filesystem.Options{
		AlternatesFS: osfs.New("/git/alternates", osfs.WithBoundOS()),
	})
	repo, err := git.Open(storer, fs)
```

**Comment by jemiahw** [comment]
I was able to get things working by using three separate filesystems. One for the storer, one for the worktree, and one for the alternates. Thank you to everyone who commented with suggestions.

My repo root is at `/builds/myproject/` and that directory has a `.git` directory as well as the working tree with all the files. Something like this:
```
/builds/myproject/.git/
/builds/myproject/.git/objects/
...
/builds/myproject/readme.md
...
```

The git repo uses an alternate. The `/builds/myproject/.git/objects/info/alternates` file points to `/mirrors/<some_temp_dir>/objects`.

To open the repo, I use code like this:
```
worktreeFS := osfs.New("/builds/myproject")
gitFS := osfs.New("/builds/myproject/.git")
altFS := osfs.New("/mirrors") 
storer = filesystem.NewStorageWithOptions(gitFS, cache.NewObjectLRUDefault(), filesystem.Options{
       AlternatesFS: altFS,
})
repo, err := git.Open(storer, worktreeFS)
```
